c++/src/capnp/serialize-async.h

```diff
@@ -139,6 +151,155 @@ class AsyncCapabilityMessageStream final: public MessageStream {
   kj::AsyncCapabilityStream& stream;
 };
 
+struct BufferedStreamOptions {
+  uint64_t defaultReadSize = 65536; // 64KiB
+  // Max amount of bytes to read in tryReadMessage if stream.tryGetLength() returns nullptr.
+
+};
+
+struct BufferedCapabilityStreamOptions : public BufferedStreamOptions {
+  uint maxFdsPerMessage;
+};
+
+class BufferedAsyncIoMessageStream;
+class BufferedAsyncCapabilityMessageStream;
+
+namespace _ {
+
+template<typename Stream>
+kj::Promise<kj::Maybe<MessageReaderAndFds>> tryReadMessageInternal(Stream& stream,
+    uint64_t defaultReadSize, kj::ArrayPtr<kj::AutoCloseFd> passedInFdSpace,
+    ReaderOptions readerOptions, kj::ArrayPtr<word> scratchSpace);
+
+extern template kj::Promise<kj::Maybe<MessageReaderAndFds>>
+      tryReadMessageInternal<BufferedAsyncIoMessageStream>(
+      BufferedAsyncIoMessageStream& stream,
+      uint64_t defaultReadSize, kj::ArrayPtr<kj::AutoCloseFd> passedInFdSpace,
+      ReaderOptions readerOptions, kj::ArrayPtr<word> scratchSpace);
+
+extern template kj::Promise<kj::Maybe<MessageReaderAndFds>>
+      tryReadMessageInternal<BufferedAsyncCapabilityMessageStream>(
+      BufferedAsyncCapabilityMessageStream& stream,
+      uint64_t defaultReadSize, kj::ArrayPtr<kj::AutoCloseFd> passedInFdSpace,
+      ReaderOptions readerOptions, kj::ArrayPtr<word> scratchSpace);
+
+} // namespace _
+
+class BufferedAsyncIoMessageStream final : public MessageStream {
```

Comment:
No I hadn't, thinking about it now it seems like a nicer approach for code complexity for sure, and makes it way easier to use this with our varying VatNetwork implementations (some don't use the MessageStream class) There's a bit of a behavior change though, but we could maintain behavior by changing how we actually use the message stream.

My PR will eagerly allocate space for each message when it reads in a whole buffer from the underlying stream. However, backing our existing `AsyncMessageReader` with a `BufferedAsyncIoStream` would do so lazily (per-call to `tryReadMessage`). This seems desirable if you're reading lots of partial messages, as it eliminates the copy to the start of the read buffer and allows you to use scratch space (right now I'm not making use of it in the buffered reader because of the eager message reading).

I think the behavior of partial message reading could be improved perhaps, but the lazy approach means that there's a good chance that subsequent messages are flushed out of processor cache if you are doing work between calls to `tryReadMessage`, which due to the `evalLast` in the rpc system message reading loop, we will. Maybe I'm wrong and this doesn't matter in practice...

However if it does matter, we could modify how `messageLoop()` works such that it eagerly reads all messages that are available and queues them before actually processing them, so there's no work happening in between reading messages which should achieve the same goal of keeping the processor cache warm. Delegating eager behavior to client code also lets it use scratch space if it wants, which seems nicer.

Reply:
Yeah I suspect that the big benefit here comes from reducing syscalls. That said, if necessary it will always be possible to optimize further by having MessageReader use dynamic_cast to detect if the underlying stream is a buffered stream, and have the buffered stream interface provide methods to query the buffer. I think that will allow lighter-touch optimizations, if they are needed at all. The non-async versions of streams have all this already, FWIW.

Reply:
There's a noticeable ~8-10% improvement in RPS for `BufferedAsyncMessageReader` over `AsyncIoMessageReader` + underlying buffered input stream -- I'll take a crack at the dynamic_cast optimization approach

Reply:
Hm, thinking about this it doesn't really make sense? `MessageReader` is per-message, so there's no way for 1 message reader to read multiple messages if it knows the underlying stream is buffered.

it seems more like the `MessageStream` classes would need to be made buffer-aware, and at that point I think the code ends up pretty similar to what I have now, except now we have a buffered async byte stream class to use elsewhere.

Reply:
How much does buffering improve performance over not buffering?

Reply:
To put it another way:

I am not very excited about reviewing this change as it seems to pile a whole lot of complexity together.

I would prefer to review a change that implements a buffered input stream class, and then possibly separate incremental changes that make MessageStream increasingly aware of buffering when it's reading from such a stream and can meaningfully optimize based on that. I think that would be easier to review and easier to maintain going forward.

Reply:
Buffering with a buffered input stream is about a 10-12% improvement in throughput measured by req/s, Buffering with my buffered message reader is a 20% improvement measured by req/s